### PR TITLE
Vectorize Hashing in FlatGroupByHash

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
Implements column-wise hash calculations in FlatHashCompiler and changes FlatGroupByHash to use it to implement a batched approach to first computing a range of position hashes and then attempting to insert those positions using the hashes that were precomputed.

By calculating position hashes in a columnar traversal, we can avoid repeated expensive bounds checking per access, allow the JIT to emit more efficient unrolled loops, and access the memory in a way that friendlier and more predictable for CPU caches.

Additionally, when attempting to insert the positions into the FlatGroupByHash after precomputing the hash code we can start loading the relevant portition of the hash table into memory sooner since we aren't intermixing computing the hash with the memory accesses.

Since re-hashing is still performed row-at-a-time, the performance improvement is much more significant when the number of groups is small and when the number of columns in the hash calculation are larger.

`BenchmarkGroupByHash.addPages` results generally improve between 5-50% depending on the scenario:
- [300 groups](https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/pettyjamesm/53cffa0222ce42afb77b3532417759ce/raw/81a3565cff65b13755c0e62709e373b6939c62e9/baseline_hash_300.json,https://gist.githubusercontent.com/pettyjamesm/53cffa0222ce42afb77b3532417759ce/raw/aea22daf843bbfe328dd0ec4d098fe872fb04ae1/vectorized_hash_300_v2.json)
- [3M groups](https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/pettyjamesm/2e3ecb34df16d34dd92ed8f70547e97f/raw/c85d5c42ff0dce774797247a013cba19ca5be58e/baseline_hash_3M.json,https://gist.githubusercontent.com/pettyjamesm/2e3ecb34df16d34dd92ed8f70547e97f/raw/96374657456986e43e2402a8436f4bdbab8fff95/vectorized_hash_3M_v2.json)

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

**Comment by wendigo** [comment]
test-other-modules failure is unrelated. I just fixed that in master. Sorry :)

**Comment by pettyjamesm** [comment]
> test-other-modules failure is unrelated. I just fixed that in master. Sorry :)

No worries, thanks for the heads up- just rebased to pick up the fix.

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java [hunk 1]
```diff
@@ -352,6 +357,85 @@ private static void generateHashBlock(ClassDefinition definition, List<KeyField>
         body.append(result.ret());
     }
 
+    private static void generateHashBlocksVectorized(ClassDefinition definition, List<KeyField> keyFields, CallSiteBinder callSiteBinder)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java:362]: I'd name this "batched" instead of "vectorized"

**Review by martint — APPROVED** [review]
Squash the first three commits, as they are all essentially incremental parts of the same change.

File: core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java]: Renamed

**Comment by mosabua** [comment]
Please note that we added a release notes entry for the performance improvement @pettyjamesm  .. for future reference and as an example.

**Pull request merged**